utils/utils.go

```diff
@@ -31,6 +32,24 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
+type ErrAgentRestart struct {
+	Msg string
+}
+
+func (e *ErrAgentRestart) Error() string {
+	return e.Msg
+}
+
+func HandleOfferingError(err error) error {
+	if err != nil {
+		var errAgentRestart *ErrAgentRestart
+		if errors.As(err, &errAgentRestart) {
+			log.Fatalf("restarting the agent %s", err)
+		}
+	}
+	return err
+}
```

Comment:
This entire error implementation does not make sense.
Please follow go type (and error) extension best practices

Reply:
- I believe we (reviewers) are offering different views to how this should be handled
- Author implemented error type but matched it wrongly at first, I told to use error value, Dhruv got rid of error altogether and you have mentioned to use error type
- However, latest comment mentions entire thing is wrong
- In this particular instance, could you pls expand on what you anticipate @nb-ohad ? (in a sense over the review period the requirement/assumptions seems to have changed as well)

below is what I'm aware wrt errors (in current context)
1. there's no need to wrap error here (`%w`)
2. error type & value is checked using `errors.As` and `errors.Is` resp. in general
3. here there's no requirement for a new error (/custom) type, let alone being in a util

I'm just sharing my opinion, happy to be corrected.

Reply:
Thank you @leelavg for noticing the different viewpoints of this specific comment. 
My original comment was to delegate a response to this class of errors - one which needs a different response than just a re-queue - to the scope outside of the offering specific code. 

The reason is that it makes sense that the response we have today (restart the operator) will change to something else (maybe, add a dynamic watch). 

My only issue with the current implementation is that it does not follow golang's best practices on type (and error) extensions.  

Saying that I do see @bindrad point that this mechanism is overcomplicated. May I suggest a shift in direction?
Maybe we want a function, somewhere common, that will handle missing watches.  The implementation for that function, for now, will be to `os.Exit` but might change as we figure out a better way. That way the handling becomes very simple, just call the `handleMissingWatch` function when a watch is missing

WDYT?

Reply:
> one which needs a different response than just a re-queue - to the scope outside of the offering specific code.
- atm, we just need a re-queue right

> maybe, add a dynamic watch
- I believe, it's going to be entirely different implementation (based on what I understand from Dhruv's findings)

> golang's best practices on type (and error) extensions.
- I maybe dragging it further, however I'm genuinely interested what's missing here.

> might change as we figure out a better way.
- atm, I can't think of better way to restart operator within itself
- anyways, the way we are doing things here doesn't restrict any new methods that'll be found later on

> That way the handling becomes very simple, just call the handleMissingWatch function when a watch is missing
- sry, I don't understand this, maybe you are looking from a different angle that we are missing, our (at-least my) understanding is, if a specific CRD is available we need to setup watch on it or else do nothing

Reply:
@leelavg 
> atm, we just need a re-queue right

No, we need a restart, that is the entire point

> I believe, it's going to be entirely different implementation (based on what I understand from Dhruv's findings)

If we figure out ho to add a watch mid reconcile. I don't believe the flow will be that different

> I maybe dragging it further, however I'm genuinely interested what's missing here.

I will try to find a reference when I have some spare time

> atm, I can't think of better way to restart operator within itself

The restart will be using `os.Exist` the issue here is who own that operation and the relationship between offering code and general control code

> anyways, the way we are doing things here doesn't restrict any new methods that'll be found later on

That is true, but in that case, lets keep it simple and implement my latest suggestion (a function call)

Reply:
> implement my latest suggestion (a function call)
- ack, so a function call which just does `os.Exit` without errors and all